### PR TITLE
Use `get_theme_file_path` function.

### PR DESCRIPTION
**Pull request by spacedmonkey**
Use `get_theme_file_path` function to stop replicated code. 

Trac ticket: https://core.trac.wordpress.org/ticket/57629

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/global-styles-and-settings.php [hunk 1]
```diff
@@ -285,13 +285,8 @@ function wp_theme_has_theme_json() {
...
 
 	// Does the theme have its own theme.json?
-	$theme_has_support = is_readable( get_stylesheet_directory() . '/theme.json' );
-
-	// Look up the parent if the child does not have a theme.json.
-	if ( ! $theme_has_support ) {
-		$theme_has_support = is_readable( get_template_directory() . '/theme.json' );
-	}
-
+	$theme_has_support = is_readable( get_theme_file_path( 'theme.json' ) );
```
Comment (hellofromtonya) [inline src/wp-includes/global-styles-and-settings.php:411]: Noting there's a filter `'theme_file_path'` in this function, meaning the returned string passes through a filter. That's good and also potentially problematic.

Why problematic? 

What if the filter changes the returned string to a different file other than `theme.json`? If that file exists, then this function will return `true` to indicate that yes the theme does support `theme.json`.

Yes, that's the developer's responsibility to ensure their callback logic works correctly.

But I'm also thinking that in this specific function, the risk is to returning and static variable caching the wrong state, making it harder for the developer to debug.

I'm thinking to keep the logic as it is. 

@felixarntz @spacedmonkey @oandregal What do you think?

File: src/wp-includes/global-styles-and-settings.php [hunk 2]
```diff
@@ -285,13 +285,8 @@ function wp_theme_has_theme_json() {
...
+
```
Comment (felixarntz) [inline src/wp-includes/global-styles-and-settings.php]: Apparently there's some whitespace here that causes a PHPCS error?

File: src/wp-includes/global-styles-and-settings.php (hunk 1 above)
Comment (felixarntz) [reply src/wp-includes/global-styles-and-settings.php:411]: @hellofromtonya I don't really see this as a concern, as the function is intended for this purpose and is called in many different places in core. If someone _wants_ to break WordPress using a filter in the wrong way, they can in 1000s of places. Even with the existing code, someone could break the logic by misusing the `stylesheet_directory` or `template_directory` filters.

At the same time, there is one other (very minor, but potentially relevant) difference in the logic: The code here uses `is_readable` while `get_theme_file_path()` uses `file_exists`. FWIW, I do not understand why the `theme.json` code uses `is_readable` in general, as it has been common in Core to simply rely on `file_exists`, and typically there shouldn't be any difference. Do you have any context on that @oandregal?

Comment (felixarntz) [inline src/wp-includes/global-styles-and-settings.php:411]: See my overarching feedback: If we go with this approach, I think we should use `file_exists` throughout.
```suggestion
	$theme_has_support = file_exists( get_theme_file_path( 'theme.json' ) );
```

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey I think we should either leave this as is and continue to rely on the `is_readable` logic, or we should consistently switch this to use `file_exists`. Right now the PR here results in a mix of both which does not make sense to me.

I don't have a strong opinion either way. If there's no particular reason why the `theme.json` related code started using `is_readable` instead of `file_exists`, I'd be onboard with changing this code to use `get_theme_file_path()`, but then also consistently use `file_exists`.

File: src/wp-includes/global-styles-and-settings.php (hunk 1 above)
Comment (spacedmonkey) [reply src/wp-includes/global-styles-and-settings.php:411]: !00% agree with @felixarntz . You can break wordpress with filters, everyone knows that, it is not a reason not to use correctly use this function here.

File: src/wp-includes/global-styles-and-settings.php (hunk 2 above)
Comment (mukeshpanchal27) [inline src/wp-includes/global-styles-and-settings.php]:
```suggestion

```

File: src/wp-includes/global-styles-and-settings.php (hunk 1 above)
Comment (oandregal) [reply src/wp-includes/global-styles-and-settings.php:411]: Thanks for raising, Tonya. I'd think this change should be fine: the consumers of the filter are responsible for making sure it works, and the cache is non-persistent (the value is calculated in every request).

> FWIW, I do not understand why the theme.json code uses is_readable in general, as it has been common in Core to simply rely on file_exists, and typically there shouldn't be any difference. Do you have any context on that @oandregal?

I see plenty of uses for both [is_readable](https://github.com/WordPress/wordpress-develop/search?q=is_readable) and [file_exists](https://github.com/WordPress/wordpress-develop/search?q=file_exists) in core.

In this case, we want to read the file later, so we check for `is_readable`. Note that this function and the function that actually reads the file from the file system use the same `is_readable` [check](https://github.com/WordPress/wordpress-develop/blob/5cf1b5ac22c58e4492ae68bac60715213da3181f/src/wp-includes/class-wp-theme-json-resolver.php#L621). If there was a reason to update this, it'll need to be updated in both places.

**Comment by spacedmonkey** [comment]
@mukeshpanchal27 @felixarntz 

I have include another commit, to also use the `theme_file_path` filter in theme json resolver. If we are going to implement using this filter in one place, it needs to be done in all. 

think of this use case. 
- Developer, has a filter that filters the location of the theme.json file. 
- Calls `wp_theme_has_theme_json`, get return true. 
- But then, `get_theme_data` in theme json resolver, use `get_file_path_from_theme` that does not respect that.

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 3]
```diff
@@ -238,9 +238,9 @@ public static function get_theme_data( $deprecated = array(), $options = array()
 		$options = wp_parse_args( $options, array( 'with_supports' => true ) );
 
 		if ( null === static::$theme || ! static::has_same_registered_blocks( 'theme' ) ) {
-			$theme_json_file = static::get_file_path_from_theme( 'theme.json' );
 			$wp_theme        = wp_get_theme();
-			if ( '' !== $theme_json_file ) {
+			$theme_json_file = $wp_theme->get_file_path( 'theme.json' );
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-theme-json-resolver.php:242]: Benefit of this, if you have child theme without theme.json, it fallbacks to parent and then does check again below.

**Comment by hellofromtonya** [comment]
Why `is_readable()` vs `file_exists()`?

>is_readable — Tells whether a file exists and is readable
>Returns true if the file or directory specified by filename exists and is readable, false otherwise.
https://www.php.net/manual/en/function.is-readable.php

>file_exists — Checks whether a file or directory exists
>Returns true if the file or directory specified by filename exists; false otherwise.
https://www.php.net/manual/en/function.file-exists.php

`is_readable()` then is an extra safety measure to ensure the file is usable for reading it into memory.

**Comment by spacedmonkey** [comment]
> `is_readable()` then is an extra safety measure to ensure the file is usable for reading it into memory.

I recommend using `is_readable`, as this what is correctly in core. 

I think using the `theme_file_path` is extremely useful for developers. 

I would like to unblock this, can I get a review of this. 

@hellofromtonya @oandregal @felixarntz

**Comment by oandregal** [comment]
My main comment is that these changes should happen in Gutenberg first. This is something I have recurrently mentioned, and it takes time and energy for everyone. I feel I sound like a broken record. I'm sorry about that. Is there something I can do to clarify this, so nobody needs to waste energy/time on this again?

**Comment by oandregal** [comment]
For full transparency: other than I already commented [back in February](https://github.com/WordPress/wordpress-develop/pull/3876#discussion_r1098474838), I don't fully understand how this is impactful (performance, users, etc.). I feel others would be better equipped than I am to help figure this out.

**Comment by spacedmonkey** [comment]
> I feel I sound like a broken record. I'm sorry about that. 

Can you point me in the direction of the documented process for this? @oandregal

**Comment by oandregal** [comment]
> Can you point me in the direction of the documented process for this?

As far as I know, it has been good etiquette among people that work together, and it has been transmitted PR to PR in comments. I have been personally doing this for months, if not a year. I know @azaozz and @hellofromtonya have been sharing this as well in other channels, they will have more context as to where.

I don't know that there is a handbook page anywhere that states this. I'm happy to help writing one down if that's useful to help us focus on solving user's problems.

File: src/wp-includes/global-styles-and-settings.php (hunk 1 above)
Comment (spacedmonkey) [reply src/wp-includes/global-styles-and-settings.php:411]: I agree with @oandregal here. Let's stick with `is_readable`.

**Comment by johnbillion** [comment]
Following on from the above, is this process documented in a handbook yet? If there are files in core that need to be modified first in Gutenberg then ideally there should be a test that fails when they're modified in a core PR.

**Review by audrasjb — APPROVED** [review]
The logic looks good to me 👍

**Review by costdev — APPROVED** [review]
Thanks for the PR @spacedmonkey! As your plan is to backport this to Gutenberg in [this PR](https://github.com/WordPress/gutenberg/pull/45831), everything looks good to me here - and nice one using `is_readable()` over `file_exists()` 👍

**Review by felixarntz — APPROVED** [review]
Thanks @spacedmonkey, LGTM!

**Pull request closed without merging**